### ISSUE TITLE
[Bug]: Unable to ignore changes to RDS Postgres minor engine version

### ISSUE THREAD

**Issue by YevheniiPokhvalii** [hashicorp/terraform-provider-aws#39579]
### Terraform Core Version

1.9.5

### AWS Provider Version

5.69.0

### Affected Resource(s)

aws_db_instance

### Expected Behavior

Terraform should ignore the minor Postgres versions upgrade when `auto_minor_version_upgrade = true`.
(`auto_minor_version_upgrade`  is true by default)

### Actual Behavior

api error InvalidParameterCombination: Cannot upgrade postgres from 13.15 to 13.13

### Terraform Configuration Files

```
resource "aws_db_instance" "db_instance" {
  identifier              = "${var.env_name[var.environment]}-db-instance"
  engine                  = "postgres"
  db_name                 = "example_db"
  engine_version          = "13.13"
}
```

### Steps to Reproduce

Run `terraform plan`

Terraform will perform the following actions:
```
  # aws_db_instance.db_instance will be updated in-place
  ~ resource "aws_db_instance" "db_instance" {
      ~ engine_version                        = "13.15" -> "13.13"
        id                                    = "db-someID"
}
```

Run `terraform apply`

```
api error InvalidParameterCombination: Cannot upgrade postgres from 13.15 to 13.13
```

### References

This patch does not seem to fix the issue with Postgres versions: https://github.com/hashicorp/terraform-provider-aws/issues/1198

Postgres versioning: https://www.postgresql.org/docs/release/

### Would you like to implement a fix?

None

**Comment by trevorrea** [comment]
Would you not use a lifecycle block for this:-

```
  lifecycle {
    ignore_changes = [
      engine_version
    ]
  }
```

**Comment by nekketsuuu** [comment]
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance#engine_version says

> If `auto_minor_version_upgrade` is enabled, you can provide a prefix of the version such as `8.0` (for `8.0.36`).

So, how about setting engine_version to `13` instead of `13.13`?

**Comment by YevheniiPokhvalii** [comment]
> https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance#engine_version says
> 
> > If `auto_minor_version_upgrade` is enabled, you can provide a prefix of the version such as `8.0` (for `8.0.36`).
> 
> So, how about setting engine_version to `13` instead of `13.13`?

I tired that. I does not work. It tries to downgrade the database to 13 in this case. You'll see something like that:

```
  # aws_db_instance.db_instance will be updated in-place
  ~ resource "aws_db_instance" "db_instance" {
      ~ engine_version                        = "13.13" -> "13"
        id                                    = "db-someID"
}
```

**Comment by nekketsuuu** [comment]
Yes... I was falling into the same situation, too. It seems that this document was added at <https://github.com/hashicorp/terraform-provider-aws/pull/3886> several years ago, but I don't know this is the intended behavior.

**Comment by HelvetiaAppDev** [comment]
I have a very similar issue. The version in my terraform config is "16" and the RDS has 16.4 (due to automatic minor upgrades).
`Cannot upgrade postgres from 16.4 to 16.3`
Somehow TF "expands" v16 to v16.3, which would be a downgrade.

**Comment by cswingler** [comment]
One interesting note:

The behavior works fine on MySQL, but postgres behaves as documented in this ticket.

Is this a matter of Postgres following semantic versioning while [MySQL is off doing its own thing](https://dev.mysql.com/doc/refman/8.0/en/which-version.html), and Terraform not handling that case?

**Comment by jfharden** [comment]
> I have a very similar issue. The version in my terraform config is "16" and the RDS has 16.4 (due to automatic minor upgrades). `Cannot upgrade postgres from 16.4 to 16.3` Somehow TF "expands" v16 to v16.3, which would be a downgrade.

We're also seeing this. We have postgres RDS 14.18, that's in the terraform code for engine_version, and in the state for engine_version and actual_engine_version; If we change the terraform code to have engine_version 14, then terraform tries to "upgrade" from 14.18 to 14.17.

We also have `auto_minor_version_udpate` set to `true`

**Comment by m9aertner** [comment]
@jfharden @HelvetiaAppDev Can the "unexpected" minor version 17 be related to what AWS terms "standard for major version 14" ? AWS (currently) shows this next to "compatible with PostgreSQL 14.17"? So when you really have 14.18 installed and just "14" in TF, the 17 is somehow pulled from AWS to form the undesired downgrade from 14.18 -> 14.17?

**Comment by YevheniiPokhvalii** [comment]
> [@jfharden](https://github.com/jfharden) [@HelvetiaAppDev](https://github.com/HelvetiaAppDev) Can the "unexpected" minor version 17 be related to what AWS terms "standard for major version 14" ? AWS (currently) shows this next to "compatible with PostgreSQL 14.17"? So when you really have 14.18 installed and just "14" in TF, the 17 is somehow pulled from AWS to form the undesired downgrade from 14.18 -> 14.17?

It's the problem with the provider itself. I think it's clearly described in the thread. The Terraform provider pulls exactly 14, if you indicate 14 or it will pull 14.17 if you indicated 14.17. It does not understand the postgres semantic versioning and that option `auto_minor_version_udpate ` does not work there.  Back in the day, they had similar issue with MySQL too and they fixed it. Now the issue is with Postgres. Versioning in MySQL and Postgres are absolutely different. Postgres uses it's own thing.

**Comment by cjsguru** [comment]
We are handling the auto minor version thingy with this other PR:
https://github.com/cargurus-hcp-terraform/policy-library-aws/pull/147
So this ticket/PR can most probably get closed.

**Comment by cjsguru** [comment]
And just in case, auto minor version should be set as FALSE.
Not sure what the "thumbs down" mean on the previous comment; you mind typing down what is meant.

**Comment by cjsguru** [comment]
> > [@jfharden](https://github.com/jfharden) [@HelvetiaAppDev](https://github.com/HelvetiaAppDev) Can the "unexpected" minor version 17 be related to what AWS terms "standard for major version 14" ? AWS (currently) shows this next to "compatible with PostgreSQL 14.17"? So when you really have 14.18 installed and just "14" in TF, the 17 is somehow pulled from AWS to form the undesired downgrade from 14.18 -> 14.17?
> 
> It's the problem with the provider itself. I think it's clearly described in the thread. The Terraform provider pulls exactly 14, if you indicate 14 or it will pull 14.17 if you indicated 14.17. It does not understand the postgres semantic versioning and that option `auto_minor_version_udpate ` does not work there. Back in the day, they had similar issue with MySQL too and they fixed it. Now the issue is with Postgres. Versioning in MySQL and Postgres are absolutely different. Postgres uses it's own thing.

This will get fixed in the module so there's no need to pass any specific PostgreSQL version

**Comment by cjsguru** [comment]
> > I have a very similar issue. The version in my terraform config is "16" and the RDS has 16.4 (due to automatic minor upgrades). `Cannot upgrade postgres from 16.4 to 16.3` Somehow TF "expands" v16 to v16.3, which would be a downgrade.
> 
> We're also seeing this. We have postgres RDS 14.18, that's in the terraform code for engine_version, and in the state for engine_version and actual_engine_version; If we change the terraform code to have engine_version 14, then terraform tries to "upgrade" from 14.18 to 14.17.
> 
> We also have `auto_minor_version_udpate` set to `true`

autominor version flag should be set to FALSE; otherwise you might experience outages out of the blue when an autoupgrade happens.
The problem with PostgreSQL versioning will be handled within a few days in the module.
For now, it is ok what you are trying to do to get unblocked.

**Comment by nekketsuuu** [comment]
I believe you might be mistaken about something. This issue is about the official AWS provider for Terraform. The pull request you mentioned is private and does not seem to be related to this repository.
